File: templates/typescript/team_source/strategy.ts
export function chooseMove(
  turn: number,
  myHistory: string,
  opponentHistory: string,
  rng: { nextInt(upperExclusive: number): number }
): string {
  const moves = ["R", "P", "S"];
  //Naive cheating to start
  let returnValue = opposite(moves[rng.nextInt(moves.length)])
  const [rCount, pCount, sCount] = countOccurence(opponentHistory)
  let rPerc = rCount/turn
  let pPerc = pCount/turn
  let sPerc = sCount/turn
  if(rPerc > .5){
    return "P"
  }
  else if(pPerc > .5) {
    return "S"
  }
  else if(sPerc > .5) {
    return "R"
  }
  if (opponentHistory.slice(turn-4).match("/(.)\\1\\1/")) {
    return opposite(opponentHistory.charAt(turn-1))
  }
  return returnValue;
}

function opposite(thrown: string
) {
  switch (thrown) {
    case "R":
      return "P"
    case "P":
      return "S"
    default:
      return "R"
  }
}

function countOccurence(str: string): [number, number, number] {
  let rCount = 0;
      let sCount = 0;
      let pCount = 0;
  for (let i = 0; i < str.length; i++) {
    if (str[i] === "R") {
      rCount++;
    } else if (str[i] === "S") {
      sCount++;
    } else {
      pCount++;
    }
  }

  return [rCount, pCount, sCount];
}

File: templates/java/team_source/Strategy.java
import java.util.random.RandomGenerator;

public final class Strategy {
    private Strategy() {}

    /** Returns one legal move using only the wrapper-provided random stream. */
    public static String chooseMove(
        int turn,
        String myHistory,
        String opponentHistory,
        RandomGenerator rng
    ) {
        if (opponentHistory.isEmpty()) {
            String[] moves = {"R", "P", "S"};
            return moves[rng.nextInt(moves.length)];
        }

        int rCount = 0;
        int pCount = 0;
        int sCount = 0;

        for (char c : opponentHistory.toCharArray()) {
            if (c == 'R') rCount++;
            else if (c == 'P') pCount++;
            else if (c == 'S') sCount++;
        }

        if (rCount >= pCount && rCount >= sCount) {
            return "P"; // Beat Rock
        } else if (pCount >= sCount) {
            return "S"; // Beat Paper
        } else {
            return "R"; // Beat Scissors
        }
    }
}
